File: models/user_model.js
import { Schema, model } from 'mongoose';
import { v1 } from 'uuid';
import { genSaltSync, hashSync } from 'bcrypt';

const userSchema = Schema({
    id: { type: String, unique: true },
    fullName: { type: String, default: '', },
    email: { type: String, unique: true, requried: true },
    password: { type: String, required: true },
    phoneNumber: { type: String, default: '', },
    address: { type: String, default: '', },
    city: { type: String, default: '', },
    state: { type: String, default: '', },
    profileProgress: { type: Number, default: 0, },
    updatedOn: { type: Date, },
    createdOn: { type: Date, }

});

userSchema.pre('save', function (next) {
    this.id = v1();
    this.updatedOn = new Date();
    this.createdOn = new Date();

    const salt = genSaltSync(8);
    const hash = hashSync(this.password, salt);
    this.password = hash;
    next(); // if next has string that mean it has error
});

userSchema.pre(['update', 'findOneAndUpdate'], function (next) {
    const update = this.getUpdate();
    delete update._id;//?
    delete update.id;

    this.updatedOn = new Date();
    next();
});

// module.exports =  mongoose.model('user', userSchema);
const userModel = model('User', userSchema);

export default userModel;

File: ecommerce_frontend/lib/data/models/user_model.dart
class UserModel {
  String? sId;
  late String fullName;
  late String email;
  late String password;
  String? phoneNumber;
  String? address;
  String? city;
  String? state;
  int? profileProgress;
  String? id;
  String? updatedOn;
  String? createdOn;

  UserModel({
    this.sId,
    required this.fullName,
    required this.email,
    required this.password,
    this.phoneNumber,
    this.address,
    this.city,
    this.state,
    this.profileProgress,
    this.id,
    this.updatedOn,
    this.createdOn,
  });

// named constructor
  UserModel.fromJson(Map<String, dynamic> json) {
    sId = json['_id'];
    fullName = json['fullName'];
    email = json['email'];
    password = json['password'];
    phoneNumber = json['phoneNumber'];
    address = json['address'];
    city = json['city'];
    state = json['state'];
    profileProgress = json['profileProgress'];
    id = json['id'];
    updatedOn = json['updatedOn'];
    createdOn = json['createdOn'];
  }

// method
  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {}; //Map<String, dynamic>();
    data['_id'] = sId;
    data['fullName'] = fullName;
    data['email'] = email;
    data['password'] = password;
    data['phoneNumber'] = phoneNumber;
    data['address'] = address;
    data['city'] = city;
    data['state'] = state;
    data['profileProgress'] = profileProgress;
    data['id'] = id;
    data['updatedOn'] = updatedOn;
    data['createdOn'] = createdOn;
    return data;
  }
}
